Constrain pan/zoom to min/max domain values

themadcreator opened the pull request:
Previously it was only possible to set the min/max
domain EXTENT. This was only respected by the total
zoom level.

For example, if min/max extent was set to [0, 10],
then it was possible to pan to [10, 20] or zoom to
[-5, 5].

This change adds domain VALUE constraints to both pan
and zoom to respect the absolute domain boundaries.

The implementation for zoom constraints must consider
that the zoom interaction may violate the constraints
for min, max, or both.

We handle each of these cases seperately:
min only: We adjust the "center" coordinate of the zoom.
max only: We adjust the "center" coordinate of the zoom.
min and max: We adjust both the "center" and "zoom" values.

This also supports partial constraints (i.e. only min
or max set).

themadcreator committed:
Constrain pan/zoom to min/max domain values

Previously it was only possible to set the min/max
domain EXTENT. This was only respected by the total
zoom level.

For example, if min/max extent was set to [0, 10],
then it was possible to pan to [10, 20] or zoom to
[-5, 5].

This change adds domain VALUE constraints to both pan
and zoom to respect the absolute domain boundaries.

The implementation for zoom constraints must consider
that the zoom interaction may violate the constraints
for min, max, or both.

We handle each of these cases seperately:
min only: We adjust the "center" coordinate of the zoom.
max only: We adjust the "center" coordinate of the zoom.
min and max: We adjust both the "center" and "zoom" values.

This also supports partial constraints (i.e. only min
or max set).

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -232,6 +255,92 @@ namespace Plottable.Interactions {
+        return { centerPoint, zoomAmount };
+      }
+
+      const scaleDomain = scale.domain();
```
nit: just for added clarity can we do `const [scaleDomainMin, scaleDomainMax] = scale.domain();`

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,117 @@ namespace Plottable.Interactions {
+     * the user cannot pan/zoom.
+     *
+     * @param {QuantitativeScale<any>} quantitativeScale The scale to query
+     * @returns {D} The minimum domain extent for the scale.
```
"The minimum domain _value_ for the scale"

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,117 @@ namespace Plottable.Interactions {
+     * the user cannot pan/zoom.
+     *
+     * @param {QuantitativeScale<any>} quantitativeScale The scale to query
+     * @param {D} minDomainExtent The minimum domain extent for the scale.
```
copy-paste bugz

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,117 @@ namespace Plottable.Interactions {
+      }
+      let maxValueForScale = this.maxDomainValue(quantitativeScale);
+      if (maxValueForScale != null && maxValueForScale.valueOf() < minDomainValue.valueOf()) {
+        throw new Error("minDomainValue must be smaller than maxDomainValue for the same Scale");
```
should we care if they're equal? that also sounds invalid to me

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,117 @@ namespace Plottable.Interactions {
+     * the user cannot pan/zoom.
+     *
+     * @param {QuantitativeScale<any>} quantitativeScale The scale to query
+     * @returns {D} The maximum domain extent for the scale.
```
"The maximum domain _value_ for the scale."

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,117 @@ namespace Plottable.Interactions {
+     * the user cannot pan/zoom.
+     *
+     * @param {QuantitativeScale<any>} quantitativeScale The scale to query
+     * @param {D} maxDomainExtent The maximum domain extent for the scale.
```
copy-paste bug

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,117 @@ namespace Plottable.Interactions {
+     * This constrains the pan/zoom interaction to show no more than the domain
+     * of the scale.
+     */
+    public constrainToScaleDomainValues() {
```
is this method used anywhere?

themadcreator replied:
it's a new public method that we'll use in the blueprint-chart to easily attach constraints using the existing scale

CalvinFernandez replied:
ah okay -- the name of the method doesn't make it super clear what it does; this is subjective of course. @hellochar thoughts on this ? does it make sense to you?

hellochar replied:
I didn't know what this method did until I read your comment. Can we just implement this on the blueprint-chart side?

CalvinFernandez replied:
I'm also for letting end-users implement this piece.

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -244,19 +353,44 @@ namespace Plottable.Interactions {
+      if (translation > 0) {
+        const bound = this.maxDomainValue(scale);
+        if (bound != null) {
+          const currentRange = scale.scale(scaleDomain[1]);
```
nit `currentMaxRange`

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -244,19 +353,44 @@ namespace Plottable.Interactions {
+        if (bound != null) {
+          const currentRange = scale.scale(scaleDomain[1]);
+          const maxRange = scale.scale(bound);
+          translation = Math.min(currentRange + translation, maxRange) - currentRange;
```
this is so clean 👍

CalvinFernandez commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -232,6 +255,92 @@ namespace Plottable.Interactions {
+
+        const minRange = scale.scale(minDomain);
+        const currentMinRange = scale.scale(scaleDomain[0]);
+        const testMinRange = this._zoomAt(currentMinRange, zoomAmount, centerPoint);
```
could probably DRY this if you're feeling particularly motivated

themadcreator replied:
Yeah, there is some duplication here, but I couldn't think of a way to do that without losing some of the clarity of flow of control especially with all the null checking required for each of these cases. Also these calculations are super fast, so it's not a perf concern.

CalvinFernandez reviewed (commented):
looking good! needs some tests I think

themadcreator committed:
Add tests

Fix constrained pinch calculation.

Add a bunch of tests for min/max/min+max domain value constraints.

themadcreator committed:
Address PR Comments

Fix copy-paste bugs
A little more DRY
Test for minDomainValue == maxDomainValue

hellochar commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -232,6 +255,92 @@ namespace Plottable.Interactions {
       return compareF(zoomAmount, boundingDomainExtent / domainExtent);
     }
 
+    private _constrainZoomValues(scale: QuantitativeScale<number>, zoomAmount: number, centerPoint: number) {
```
I didn't think through the math here but just want to make sure this will work for logarithmic scales as well?

hellochar commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,112 @@ namespace Plottable.Interactions {
+      const [ domainMin, domainMax ] = scale.domain();
+      this.minDomainValue(scale, domainMin);
+      this.maxDomainValue(scale, domainMax);
+    }
```
this file is getting really large - over 700 lines. We should refactor this soon (would be good to revisit the codebase as a whole here)

themadcreator replied:
your point about constrainToScaleDomainValues, I think we should keep it since it's bundles up a few lines of functionality. At the very least, we should have a method to simultaneously set the min/max.

Otherwise, we can get into situations where the domain goes from [0, 100] -> [400, 500] and the first call to `minDomainValue(scale, 400)` would fail because `maxDomainValue(scale)` is still set to 100

hellochar commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,112 @@ namespace Plottable.Interactions {
+     * @param {QuantitativeScale<any>} quantitativeScale The scale to query
+     * @returns {D} The minimum domain value for the scale.
+     */
+    public minDomainValue<D>(quantitativeScale: QuantitativeScale<D>): D;
```
Is it worth calling out that the user must first `.add` the scale?

themadcreator replied:
As with minDomainExtent/maxDomainExtent, you don't actually have to add the scale to the interaction before this. You could also add it after.

hellochar commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,112 @@ namespace Plottable.Interactions {
+        return this._maxDomainValues.get(quantitativeScale);
+      }
+      let minValueForScale = this.minDomainValue(quantitativeScale);
+      if (minValueForScale != null && maxDomainValue.valueOf() <= minValueForScale.valueOf()) {
```
why do we `.valueOf`?

themadcreator replied:
I don't know, I was following the maxDomainExtent function as a guide. Figure I should be consistent with the code I find there since I'm still getting to know it.

hellochar commented on src/interactions/panZoomInteraction.ts:
```diff
@@ -439,6 +577,112 @@ namespace Plottable.Interactions {
+     * @returns {Interactions.PanZoom} The calling PanZoom Interaction.
+     */
+    public maxDomainValue<D>(quantitativeScale: QuantitativeScale<D>, maxDomainValue: D): this;
+    public maxDomainValue<D>(quantitativeScale: QuantitativeScale<D>, maxDomainValue?: D): any {
```
nit: return type should be `D | this`

Aside: how do you guys feel about this fluent method style for settable properties? The method naming feels confusing to me. Perhaps changing these to getter/setters would help differentiate between "properties that you can set" and "methods that act" (e.g. `attachTo` or `computeLayout`)?

themadcreator replied:
Okay, the type is easy enough to change.

As to your other comment, if we really wanted to differentiate these, why not use actual getter/setter properties? Obviously it exposes the possibility of side effects when doing things as simple assignment. Nevertheless, it seems like a more modern approach.

hellochar reviewed (commented):
Looks fine; would like to check log scales just to ensure this works before we merge.

themadcreator committed:
Proper return typing

themadcreator committed:
Rename constrain method

hellochar reviewed (commented):
Did we test this with log scales? Looks fine otherwise

hellochar reviewed (changes requested):
Oh sorry @themadcreator you'll need to add the built plottable.js and plottable-npm.d.ts files as well (see https://github.com/palantir/plottable#contributing)

themadcreator committed:
Add built artifacts
